cudax/include/cuda/experimental/__memory_resource/cuda_async_memory_resource.h

```diff
@@ -0,0 +1,294 @@
+
+    ::cudaMemPool_t __pool;
+    _CCCL_TRY_CUDA_API(
+      ::cudaDeviceGetDefaultMemPool, "Failed to call cudaDeviceGetDefaultMemPool", &__pool, __device_id);
+    return __pool;
+  }
+
+public:
+  //! @brief Default constructs the cuda_async_memory_resource using the default \c cudaMemPool_t of the default device
+  //! @throws cuda_error if retrieving the default \c cudaMemPool_t fails
+  cuda_async_memory_resource()
+      : __pool_(__get_default_mem_pool(0))
+  {}
+
+  //! @brief Constructs a cuda_async_memory_resource using the default \c cudaMemPool_t of a given device
+  //! @throws cuda_error if retrieving the default \c cudaMemPool_t fails
+  cuda_async_memory_resource(::cuda::experimental::device_ref __device)
+      : __pool_(__get_default_mem_pool(__device.get()))
+  {}
+
+  cuda_async_memory_resource(int)                   = delete;
+  cuda_async_memory_resource(_CUDA_VSTD::nullptr_t) = delete;
+
+  //! @brief  Constructs the cuda_async_memory_resource from a \c cudaMemPool_t.
+  //! @param __pool The \c cudaMemPool_t used to allocate memory.
+  explicit cuda_async_memory_resource(::cudaMemPool_t __pool) noexcept
+      : __pool_(__pool)
+  {}
+
+  //! @brief  Constructs the cuda_async_memory_resource from a \c cuda_memory_pool by calling native_handle()
+  //! @param __pool The \c cuda_memory_pool used to allocate memory.
+  explicit cuda_async_memory_resource(cuda_memory_pool& __pool) noexcept
+      : __pool_(__pool.native_handle())
+  {}
+
+  //! @brief Allocate device memory of size at least \p __bytes via cudaMallocFromPoolAsync.
+  //! @param __bytes The size in bytes of the allocation.
+  //! @param __alignment The requested alignment of the allocation.
+  //! @throws cuda::std::bad_alloc In case of invalid alignment, or
+  //! @throws cuda::cuda_error If an error code was return by the cuda api call.
+  //! @return Pointer to the newly allocated memory
+  _CCCL_NODISCARD void* allocate(const size_t __bytes,
+                                 const size_t __alignment = _CUDA_VMR::default_cuda_malloc_alignment)
+  {
+    if (!__is_valid_alignment(__alignment))
+    {
+      _CUDA_VSTD_NOVERSION::__throw_bad_alloc();
+    }
+
+    void* __ptr{nullptr};
+    _CCCL_TRY_CUDA_API(
+      ::cudaMallocFromPoolAsync,
+      "cuda_async_memory_resource::allocate failed to allocate with cudaMallocFromPoolAsync",
+      &__ptr,
+      __bytes,
+      __pool_,
+      ::cudaStream_t{0});
+    return __ptr;
+  }
+
+  //! @brief Deallocate memory pointed to by \p __ptr.
+  //! @param __ptr Pointer to be deallocated. Must have been allocated through a call to `allocate`
+  //! @param __bytes  The number of bytes that was passed to the `allocate` call that returned \p __ptr.
+  //! @param __alignment The alignment that was passed to the `allocate` call that returned \p __ptr.
+  void deallocate(void* __ptr, const size_t, const size_t __alignment = _CUDA_VMR::default_cuda_malloc_alignment)
```

Comment:
I would say yes

Reply:
In deallocate you can `assert` that there are no CUDA errors. (Also catch any CUDA exceptions). In allocate, throw an exception if there are CUDA errors.  Since deallocate is very commonly called in destructors, it should not throw.